### PR TITLE
Make regex a define exclusive

### PR DESCRIPTION
**Pull request by Bmooij**
Make regex a define exclusive

Based on feedback: https://github.com/me-no-dev/ESPAsyncWebServer/pull/615#issuecomment-538722184
And: https://github.com/me-no-dev/ESPAsyncWebServer/issues/613

It is not finished/tested yet, but feedback is always welcome.

### CODE REVIEW

**Comment by me-no-dev** [comment]
maybe have the public functions do nothing or return error if not enabled. and why is the define name change necessary?

**Comment by me-no-dev** [comment]
checks are failing in the regex example

**Comment by Bmooij** [comment]
> maybe have the public functions do nothing or return error if not enabled. 
I was struggling with that. 
On one side for code readability, I should say "return error if not enabled". (Run time error)
On the other side, if not enabled, you cannot use the corresponding functions. (Compile time error)

> why is the define name change necessary?
I was not sure of the name of the define.
So there is one place to change the name of the define.

**Comment by Bmooij** [comment]
I reverted the performance improvement. 
Mentioned [here](https://github.com/me-no-dev/ESPAsyncWebServer/issues/613#issuecomment-537918572)

It uses (a lot) more heap, for to little performance  improvement (in my opinion)

File: .github/scripts/install-arduino-ide.sh [hunk 1]
```diff
@@ -107,6 +108,7 @@ function build_sketch(){ # build_sketch <fqbn> <path-to-ino> [extra-options]
 		-libraries "$ARDUINO_USR_PATH/libraries" \
 		-build-cache "$ARDUINO_CACHE_DIR" \
 		-build-path "$ARDUINO_BUILD_DIR" \
+		-prefs=compiler.cpp.extra_flags="$build_flags" \
```
Comment (me-no-dev) [inline .github/scripts/install-arduino-ide.sh]: this should maybe include the c flags as well, not just cpp

File: .github/scripts/install-arduino-ide.sh [hunk 2]
```diff
@@ -210,7 +212,13 @@ function build_sketches() # build_sketches <fqbn> <examples-path> <chunk> <total
         || [ "$sketchnum" -gt "$end_index" ]; then
         	continue
         fi
-        build_sketch "$fqbn" "$sketch" "$xtra_opts"
+		local sketchBuildFlags=""
+		if [ -f "$sketchdir/.test.build_flags" ]; then
+			while read line; do
+				sketchBuildFlags="$sketchBuildFlags $line"
+			done < "$sketchdir/.test.build_flags"
+		fi
```
Comment (me-no-dev) [inline .github/scripts/install-arduino-ide.sh]: please care for tabs/spaces. in other words... do not use tabs :)

File: README.md [hunk 3]
```diff
@@ -1484,3 +1485,30 @@ webServer.onNotFound([](AsyncWebServerRequest *request) {
...
+```cpp
+  server.on("^\\/sensor\\/([0-9]+)$", HTTP_GET, [] (AsyncWebServerRequest *request) {
+      String sensorId = request->pathArg(0);
+  });
+```
+*NOTE*: All regex patterns starts with `^` and ends with `$`
+
+To enable the `Path variable` support, you have to define the buildflag `-DASYNCWEBSERVER_REGEX`.
+
+For Arduino IDE modify `platform.local.txt`:
```
Comment (me-no-dev) [inline README.md]: where is that platform.local.txt?

File: src/ESPAsyncWebServer.h [hunk 4]
```diff
@@ -271,13 +273,20 @@ class AsyncWebServerRequest {
...
     const String& headerName(size_t i) const;    // get request header name by number
     String urlDecode(const String& text) const;
+
+#ifdef _ASYNCWEBSERVER_REGEX
+    private:
+      LinkedList<String *> _pathParams;
+      void _addPathParam(const char *param);
+
+    public:
+      const String& pathArg(size_t i) const;
```
Comment (me-no-dev) [inline src/ESPAsyncWebServer.h]: this functions should exists either way. just return whatever is the equivalent of error when regex is disabled.

Comment (me-no-dev) [reply src/ESPAsyncWebServer.h]: in other words... examples that use it, should compile regardless if regex is enabled or not.

Comment (Bmooij) [reply src/ESPAsyncWebServer.h]: Personally I'm not a fan of run time errors. Now you get a error on compile time when ASYNCWEBSERVER_REGEX was not defined.

But then again that is my opinion.

Comment (Bmooij) [reply src/ESPAsyncWebServer.h]: I've changed it. If ASYNCWEBSERVER_REGEX is not defined it handles the function if the pathArg was not found.

I added a compiler warning, if the function is used but ASYNCWEBSERVER_REGEX was not defined

File: README.md (hunk 3 above)
Comment (Bmooij) [reply README.md]: I found out that platform.local.txt is not needed.

For ardruino IDE:
```cpp
#define ASYNC_WEBSERVER_REGEX
#include <ESPAsyncWebServer.h>
```

File: .github/scripts/install-arduino-ide.sh (hunk 2 above)
Comment (Bmooij) [reply .github/scripts/install-arduino-ide.sh]: Replaced all the tabs with spaces

File: .github/scripts/install-arduino-ide.sh (hunk 1 above)
Comment (Bmooij) [reply .github/scripts/install-arduino-ide.sh]: Currently I'm only using a CPPDEFINE

**Comment by Bmooij** [comment]
> why change all tabs/spaces. change ONLY your edits. now I have no clue what is going on. Revert that and apply only on your changed lines.

Sorry for the inconvenience, I reverted it and applied it only on my added lines

File: README.md (hunk 3 above)
Comment (Bmooij) [reply README.md]: I will switch back to the platform.local.txt method. 
I do not prefer that method, but it is the only method that works with the Arduino IDE

Comment (me-no-dev) [reply README.md]: yes, unfortunately we can not add menus from the libraries (or something like that). Arduino is quite more basic that most of the editors that I use. Users know to poke around. That is how ESP8266 has SSL support in ESPAsyncTCP

**Comment by me-no-dev** [comment]
alright :) it looks really nice now. Thanks for adding docs ;) Will merge when the build finishes :)

**Pull request merged**